File: src/strategy.js
import { createRegexStrategy } from "./strategies/regex.js";
import { createTrieStrategy } from "./strategies/trie.js";

/**
 * Number of dynamic routes at or above which the trie strategy is selected
 * over the regex strategy. Considered an implementation detail.
 *
 * @type {number}
 */
export const DYNAMIC_ROUTE_THRESHOLD = 50;

/**
 * Select the most appropriate matching strategy for a compiled collection.
 *
 * The decision is automatic and based on the number of dynamic routes. A
 * small collection uses the regex strategy; a large collection uses the trie
 * strategy for better scaling. The public API and matching behavior remain
 * identical regardless of the chosen strategy.
 *
 * @param {import("./compile.js").CompiledRoutes} compiled - The compiled route collection.
 * @returns {import("./compile.js").MatchStrategy} The selected strategy.
 */
export function selectStrategy(compiled) {
  if (compiled.dynamicRoutes.length >= DYNAMIC_ROUTE_THRESHOLD) {
    return createTrieStrategy(compiled);
  }
  return createRegexStrategy(compiled);
}

File: go/strategy.go
package roution

// DynamicRouteThreshold is the number of dynamic routes at or above which the
// "auto" strategy selects the trie implementation. Considered an
// implementation detail.
const DynamicRouteThreshold = 50

// SelectStrategy chooses the most appropriate matcher implementation for a
// compiled route collection.
//
// When Strategy is "auto" (the default), the decision is automatic and based on
// the number of dynamic routes: a small collection uses the regex strategy and
// a large collection uses the trie strategy. The explicit "regex" and "trie"
// values bypass the automatic decision. The public API and matching behavior
// remain identical regardless of the chosen strategy.
func SelectStrategy(compiled CompiledRoutes, opts Options) Matcher {
	switch opts.Strategy {
	case StrategyTrie:
		return NewTrieMatcher(compiled, opts.Query)
	case StrategyRegex:
		return NewRegexMatcher(compiled, opts.Query)
	default:
		if len(compiled.DynamicRoutes) >= opts.DynamicThreshold {
			return NewTrieMatcher(compiled, opts.Query)
		}
		return NewRegexMatcher(compiled, opts.Query)
	}
}
